### ISSUE TITLE
Map geometry issue

### ISSUE THREAD

**Issue by mstenta** [farmOS/farmOS#293]
Reported by @botlfarm in the forum:

> I’m having an issue where I cant assigning geometry to a log when I and an asset to it. If I dont add an asset the standard geometry map tools show up.
> 
> For example:
> scenario 1: I create an activity log, go to location/geometry (or assets/movement/geometry, with no asset selected) and the map zooms in and I have the geometry tools on the map. This appears to work normal.
> 
> scenario 2: I create an activity log, add an asset, go to asset/movement/geometry or (location/geometry) and the map stays at the world view and I have no edit tools.
> 
> scenario 3: I create an activity log, open up assets/movement/geometries and create a geometry with the mapping tools. Then I add an asset and all the geometries disappear and the mapping tools go away.
> 
> scenario 4: I create an activity log with an asset and no geometries, and save it. Then go back and edit it I can change the geometry as expected with no issues.
> 
> For background this was happening in 1.3 and after an upgrade to 1.4 is still the same.
> 
> Any thoughts on what is going on here?

https://farmos.discourse.group/t/map-geometry-issue/367

**Comment by mstenta** [comment]
@botlfarm I think I experienced this too at one point, and forgot to dig into it. It may be a JavaScript issue. My hunch is that some of the new map JS is not using Drupal JS Behaviors correctly, and so AJAX requests are breaking it. AJAX is fired when you select assets.

Can you check to see if there are any JavaScript errors in your browser console when you do this? Hopefully this will be easy to fix. I wish I had caught it before releasing 7.x-1.4... oh well. :-)

**Comment by mstenta** [comment]
My hunch is that it's a similar issue to this one: https://github.com/farmOS/farmOS/pull/275

**Comment by botlfarm** [comment]
When I click on add asset I get this error in console. 

`[Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
  | send | @ | jquery.min.js?v=2.1.4:4
-- | -- | -- | --
  | ajax | @ | jquery.min.js?v=2.1.4:4
  | n._evalUrl | @ | jquery.min.js?v=2.1.4:4
  | domManip | @ | jquery.min.js?v=2.1.4:3
  | prepend | @ | jquery.min.js?v=2.1.4:3
  | insert | @ | ajax.js?v=7.70:546
  | Drupal.ajax.success | @ | ajax.js?v=7.70:428
  | success | @ | ajax.js?v=7.70:189
  | b.success | @ | jquery.form.min.js?v=4.2.1:11
  | j | @ | jquery.min.js?v=2.1.4:2
  | fireWith | @ | jquery.min.js?v=2.1.4:2
  | x | @ | jquery.min.js?v=2.1.4:4
  | (anonymous) | @ | jquery.min.js?v=2.1.4:4
  | load (async) |   |  
  | send | @ | jquery.min.js?v=2.1.4:4
  | ajax | @ | jquery.min.js?v=2.1.4:4
  | j | @ | jquery.form.min.js?v=4.2.1:11
  | a.fn.ajaxSubmit | @ | jquery.form.min.js?v=4.2.1:11
  | Drupal.ajax.eventResponse | @ | ajax.js?v=7.70:279
  | (anonymous) | @ | ajax.js?v=7.70:206
  | dispatch | @ | jquery.min.js?v=2.1.4:3
  | r.handle | @ | jquery.min.js?v=2.1.4:3

`

**Comment by mstenta** [comment]
Thanks for the details @botlfarm ! Going to look into this...

The `[Deprecation]` notice you're seeing in console is unrelated, I believe.

I'm able to replicate these issues and I'm not seeing any useful stuff in console. Oh well.

Worth noting, I stumbled upon this same issue earlier in the Soil Test log type. If you open the "Sample Points" map, but then click the "Look up soil names" button, the map goes blank. It appears to be the exact same issue. In your case, you found that clicking the "Add items" in the "Assets" section does the same thing.

I'll let you know if I learn anything...

**Comment by mstenta** [comment]
OK I see what is causing this...

https://github.com/farmOS/farmOS/blob/7d9e6bb4a19186c897752b1687b2d10a54ef35fa/modules/farm/farm_map/js/farm_map.js#L22

The main Drupal JS behavior for creating/activating the farmOS-map instance on the page implements both an `attach()` method (for running `farmOS.map.create()`), and a `detach()` method (for running `farmOS.map.destroy()`).

If you disable the call to `farmOS.map.destroy()`, this issue is resolved.

I am trying to remember if I added that `detach()` method intentionally to fix something specific, or if I was just trying to follow best practice for "cleaning up" the behavior. To be honest, this might be the first time I've ever implemented a `detach()` JS behavior method. Maybe I misunderstood their purpose.

This doc has some good info, but I still don't feel like I have a great sense of what we need to do: https://www.drupal.org/docs/7/converting-drupal-6-modules-to-drupal-7/converting-6x-modules-to-7x#drupal_behaviors

**Comment by mstenta** [comment]
I remember why I added the `detach()` behavior. Without it, whenever AJAX/AHAH replaces a map on the page (eg: when you import a KML file), it creates a new map instance in `window.farmOS.map.instances`. The `detach()` method cleans that up.

So the question is: how do we ensure that `detach()` runs when it needs to, but not when it doesn't?

**Comment by mstenta** [comment]
OK, in the interest of fixing the more urgent bugs (described above), I decided to remove the `detach()` method, and open a new bug report specifically for the `farmOS.map.instances` cleanup issue. This is a very minor issue, by comparison, so I felt OK introducing this regression and punting it for a future fix. It should not be an issue in practice.

Here is the commit that removes the method, and fixes this issue: https://github.com/farmOS/farmOS/commit/be723c11494ccb6df878367ef57a48ddb574d52e

Here is the new bug report I opened for the regression: https://www.drupal.org/project/farm/issues/3144530

Thanks again for reporting @botlfarm - glad we could get this fixed!
